File: src/main/java/net/flyingfishflash/loremlist/core/response/structure/ApiProblemDetail.java
package net.flyingfishflash.loremlist.core.response.structure;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.databind.JsonNode;
import java.util.List;
import java.util.Map;
import org.springframework.http.ProblemDetail;

/** RFC 9457 - Problem Details for HTTP APIs */
@JsonInclude(JsonInclude.Include.NON_EMPTY)
public record ApiProblemDetail(
    String type,
    String title,
    int status,
    String detail,
    ExceptionCauseDetail cause,
    List<String> validationErrors,
    Map<String, JsonNode> supplemental,
    List<String> stackTrace) {

  public ApiProblemDetail(String type, String title, int status, String detail) {
    this(type, title, status, detail, null, null, null, null);
  }

  /** Construct an ApiProblemDetail from a Spring ProblemDetail */
  public ApiProblemDetail(ProblemDetail problemDetail) {
    this(
        problemDetail.getType().toString(),
        problemDetail.getTitle() != null ? problemDetail.getTitle() : "default title",
        problemDetail.getStatus(),
        problemDetail.getDetail() != null ? problemDetail.getDetail() : "default detail");
  }
}

File: src/main/kotlin/net/flyingfishflash/loremlist/core/response/structure/ApiProblemDetail.kt
package net.flyingfishflash.loremlist.core.response.structure

import com.fasterxml.jackson.annotation.JsonInclude
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement
import org.springframework.http.ProblemDetail

/** RFC 9457 - Problem Details for HTTP APIs **/
@Serializable
@JsonInclude(JsonInclude.Include.NON_EMPTY)
data class ApiProblemDetail(
  val type: String,
  val title: String,
  val status: Int,
  val detail: String?,
  val cause: ExceptionCauseDetail? = null,
  val validationErrors: List<String>? = null,
  val supplemental: Map<String, JsonElement>? = null,
  val stackTrace: List<String>? = null,
) {
  /** Construct an ApiProblemDetail from a Spring ProblemDetail */
  constructor(problemDetail: ProblemDetail) : this(
    type = problemDetail.type.toString(),
    title = problemDetail.title ?: "default title",
    status = problemDetail.status,
    detail = problemDetail.detail ?: "default detail",
  )
}

/** Root cause of the exception, if applicable */
@Serializable
data class ExceptionCauseDetail(
  val name: String,
  val message: String?,
)
